[JIT] improve alias analysis for list constructs

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39111 [JIT] improve alias analysis for list constructs**

In our present alias analysis, we consider any Value that enter another container as entering the heap, and thus aliasing all other heap values of the same type. There are a number of advantages to this approach:
- it is not to hard to maintain the aliasDb implementation 
- it is much easier from an op schema perspective - there are many composite list ops registered internally and externally that would be tricky to register and get right if we did something more complicated
- It limits the size of the AliasDb, because a container of size 10 only contains a single memory dag element instead of 10 elements. 

The downside is that we have are unable to handle the simple and extremely common case of a list of tensors being used in an ATen op. 

In an example like:

```
 def foo(input):
    x = torch.tensor([1, 2, 3, 4])
    y = [x, x]
    input.add_(1)
    return torch.cat(y)
```

we will consider x to be written to. any write to any wildcard element (an element that enters a tuple, an element that is taken from a list) will mark x as written to. This can be limiting for our ability to create a functional subset and fuse graphs - as a result, 4 of TorchVision classification models could not be functionalized.

Differential Revision: [D23828003](https://our.internmc.facebook.com/intern/diff/D23828003)

Commit:
[JIT] improve alias analysis for list constructs

[ghstack-poisoned]

Commit:
Update on "[JIT] improve alias analysis for list constructs"


In our present alias analysis, we consider any Value that enter another container as entering the heap, and thus aliasing all other heap values of the same type. There are a number of advantages to this approach:
- it is not to hard to maintain the aliasDb implementation 
- it is much easier from an op schema perspective - there are many composite list ops registered internally and externally that would be tricky to register and get right if we did something more complicated
- It limits the size of the AliasDb, because a container of size 10 only contains a single memory dag element instead of 10 elements. 

The downside is that we have are unable to handle the simple and extremely common case of a list of tensors being used in an ATen op. 

In an example like:

```
 def foo(input):
    x = torch.tensor([1, 2, 3, 4])
    y = [x, x]
    input.add_(1)
    return torch.cat(y)
```

we will consider x to be written to. any write to any wildcard element (an element that enters a tuple, an element that is taken from a list) will mark x as written to. This can be limiting for our ability to create a functional subset and fuse graphs - as a result, 4 of TorchVision classification models could not be functionalized. 

In subsequent PRs I can handle other common use cases such as align_tensors and broadcast_tensors - ops which take in a tensor list and return a tensor list. 






[ghstack-poisoned]

suo commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -878,6 +874,44 @@ void AliasDb::analyzeConservative(Node* node) {
+  common ops where the output does not alias the list or the list elements
+  */
+
+  switch (use.user->kind()) {
```
What are your thoughts on how we should handle these in the long run? Ideally, this pass doesn't know much about specific ops (besides language primitives) and there is a structured way (alias annotations) for the builtin operator library to communicate aliasing/mutation. Do we feel like aliasing annotations are insufficiently expressive for what we want? Too hard to change?

eellison replied:
It's certainly possible for this to be supported in the future. One difficulty is that this is backwards-incompatible change; a schema like `aten::dict((Tensor, Tensor)[] inputs -> Dict(Tensor, Tensor)` doesn't currently annotate anything about the alias set of the contained types. We don't need to specify that the keys and values of the input list now are respectively in the keys and values of the Dict type because they were all already in the wildcard set to begin with. Maybe something like:
`aten::dict((Tensor(a->c), Tensor(b->d)[] inputs -> Dict(Tensor(c), Tensor(d))` or just `aten::dict((Tensor(a), Tensor(b))[] inputs -> Dict(Tensor(a), Tensor(b))`. I don't think this is necessarily blocking, it's just something we would have to consider... There are not that many of these ops, especially in userland. 

There are other implementation difficulties as well, and potentially compilation regressions. I remember something Densenet has a list which has 64 different append nodes, so now the contained elements go from 1 -> 64. 

The maintainability and consistency of alias analysis is obviously important, but at the end of a day it is a tool that enables other optimization, and not a goal in and of itself. I would generally like to avoid sinking too much time into improving alias analysis if doesn't buy us anything.

eellison replied:
@suo wdyt?

suo commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -1068,7 +1116,9 @@ void AliasDb::createValue(const Value* value) {
-void AliasDb::giveFreshAlias(const Value* value) {
+void AliasDb::giveFreshAlias(
+    const Value* value,
+    bool add_wildcard_to_contained_elems) {
```
Seems like sort of a smell that this method grew this flag. It's not clear from the signature alone when it would be appropriate to set it true or false, since it depends heavily on callsite-specific conditions. In general, it is a sign the code is factored oddly when there is a flag to a method which is set for only a single callsite, whereas every other callsite uses a different behavior.

Ideally, the aliasing language encodes these relationships so we don't need to special case the code for it, right?

eellison replied:
yea, agreed it isn't ideal, see comment above for general discussion of aliasing language respond here for code-specific suggestions/refactorings

suo reviewed (approved):
Accepting to unblock, but I think directionally this patch raises issues about how far we want to push with special-casing patterns in alias analysis.

Commit:
Update on "[JIT] improve alias analysis for list constructs"


In our present alias analysis, we consider any Value that enter another container as entering the heap, and thus aliasing all other heap values of the same type. There are a number of advantages to this approach:
- it is not to hard to maintain the aliasDb implementation 
- it is much easier from an op schema perspective - there are many composite list ops registered internally and externally that would be tricky to register and get right if we did something more complicated
- It limits the size of the AliasDb, because a container of size 10 only contains a single memory dag element instead of 10 elements. 

The downside is that we have are unable to handle the simple and extremely common case of a list of tensors being used in an ATen op. 

In an example like:

```
 def foo(input):
    x = torch.tensor([1, 2, 3, 4])
    y = [x, x]
    input.add_(1)
    return torch.cat(y)
```

we will consider x to be written to. any write to any wildcard element (an element that enters a tuple, an element that is taken from a list) will mark x as written to. This can be limiting for our ability to create a functional subset and fuse graphs - as a result, 4 of TorchVision classification models could not be functionalized. 

[ghstack-poisoned]
